src/main/java/com/uber/cadence/internal/replay/DecisionsHelper.java

```diff
@@ -588,6 +620,57 @@ private void addDecision(DecisionId decisionId, DecisionStateMachine decision) {
     nextDecisionEventId++;
   }
 
+  // This is to support the case where a getVersion call presents during workflow execution but
+  // is removed in replay.
+  void addAllMissingVersionMarker(
+      boolean isNextDecisionVersionMarker, Optional<Predicate<byte[]>> isDifferentChange) {
+    boolean added;
+    do {
+      added = addMissingVersionMarker(isNextDecisionVersionMarker, isDifferentChange);
+    } while (added);
+  }
+
+  private boolean addMissingVersionMarker(
+      boolean isNextDecisionVersionMarker, Optional<Predicate<byte[]>> changeIdEquals) {
+    Optional<HistoryEvent> optionalEvent = getOptionalDecisionEvent(nextDecisionEventId);
+    if (!optionalEvent.isPresent()) {
+      return false;
+    }
+
+    HistoryEvent event = optionalEvent.get();
+    if (event.getEventType() != EventType.MarkerRecorded) {
+      return false;
+    }
+
+    if (!event
+        .getMarkerRecordedEventAttributes()
+        .getMarkerName()
+        .equals(ClockDecisionContext.VERSION_MARKER_NAME)) {
+      return false;
+    }
+
+    // If we have a version marker in history event but not in decisions, let's add one.
+    if (!isNextDecisionVersionMarker
+        || (changeIdEquals.isPresent()
+            && changeIdEquals.get().test(event.getMarkerRecordedEventAttributes().getDetails()))) {
```

Comment:
Make event.getMarkerRecordedEventAttributes() a local variable as it is called multiple times within this method